buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/license/RemoteLicenseFetcher.groovy

```diff
@@ -14,155 +14,159 @@
+  /**
+   * Downloads and returns the text of the license.
+   *
+   * <p>This method will try at most three times to download the license. Failures will be silently
+   * ignored if the license can be successfully downloaded on the second or third attempt. This
+   * method performs a simple, linear backoff in the case of failures to improve chances of
+   * successful connections.
+   */
+  public final String getText() {
+    IOException storedEx = null
+
+    for (int i = 0; i < 3; ++i) {
+      try {
+        if (i > 0) {
+          Thread.sleep(i * 1000)
+        }
+
+        return processDocument(Jsoup.connect(remoteUrl).get())
+      } catch (IOException ex) {
+        if (storedEx == null) {
+          storedEx = ex
+        } else {
+          storedEx.addSuppressed(ex)
+        }
+      }
+    }
+
+    throw storedEx
+  }
 
-      TEXT_FORMATTER.getPlainText(doc.select('#body-content > div.jd-descr > div')[0])
-    }
+  /** Extracts the license text from the rest of the document. */
+  String processDocument(Document doc) {
+    return TEXT_FORMATTER.getPlainText(doc)
   }
 
-  static final class Apache2LicenseFetcher implements RemoteLicenseFetcher {
-    private URI APACHE_2_LICENSE_URI = URI.create("http://www.apache.org/licenses/LICENSE-2.0.txt")
+  static final class AndroidSdkTermsFetcher extends RemoteLicenseFetcher {
 
-    @Override
-    URI getServiceUri() {
-      APACHE_2_LICENSE_URI
+    AndroidSdkTermsFetcher() {
+      super("https://developer.android.com/studio/terms.html")
     }
 
     @Override
-    String get() {
-      APACHE_2_LICENSE_URI.toURL().getText()
+    String processDocument(Document doc) {
+      // TODO(vkryachko, allisonbm92): This fails silently.
+      return ""
+      //return TEXT_FORMATTER.getPlainText(doc.select("#body-content > div.jd-descr > div"))
     }
   }
 
-  static final class AnotherApache2LicenseFetcher implements RemoteLicenseFetcher {
-    private URI APACHE_2_LICENSE_URI = URI.create("https://opensource.org/licenses/Apache-2.0")
+  static final class Apache2LicenseFetcher extends RemoteLicenseFetcher {
 
-    @Override
-    URI getServiceUri() {
-      APACHE_2_LICENSE_URI
+    Apache2LicenseFetcher() {
+      super("http://www.apache.org/licenses/LICENSE-2.0.txt")
     }
+  }
 
-    @Override
-    String get() {
-      def doc = Jsoup.connect(APACHE_2_LICENSE_URI.toString()).get()
+  static final class AnotherApache2LicenseFetcher extends RemoteLicenseFetcher {
 
-      TEXT_FORMATTER.getPlainText(doc.select('#content-wrapper'))
+    AnotherApache2LicenseFetcher() {
+      super("https://opensource.org/licenses/Apache-2.0")
     }
-  }
-
-  static final class YetAnotherApache2LicenseFetcher implements RemoteLicenseFetcher {
-    private URI APACHE_2_LICENSE_URI = URI.create("http://www.apache.org/licenses/LICENSE-2.0")
 
     @Override
-    URI getServiceUri() {
-      APACHE_2_LICENSE_URI
+    String processDocument(Document doc) {
```

Comment:
Looks like the difference among all `processDocument` lies essentially on the CSS selector part. Maybe just have the selector as a field so that we don't have to override `processDocument` for each fetcher?

Reply:
For the most part, that is true. However, some of the licenses are plain-text, so we skip the select method entirely. Having a method makes that a bit easier.

Reply:
+1 to @yifanyang 's comment, this is a good candidate for "prefer composition over inheritance", although to keep the code as expressive as it currently is, I'd suggest to have a member `Function<Document, String>` which could either be identity or "find by selector". so instead of this class hierarchy we end up with a list of values of the form:
```java
LicenseFetcher("http://example.com", d -> extractFrom(d));
```